mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp

```diff
@@ -3754,26 +3819,124 @@ static void initTargetDefaultBounds(
 /// only provide correct results if it's called after the body of \c targetOp
 /// has been fully generated.
 static void initTargetRuntimeBounds(
-    LLVM::ModuleTranslation &moduleTranslation, omp::TargetOp targetOp,
+    llvm::IRBuilderBase &builder, LLVM::ModuleTranslation &moduleTranslation,
+    omp::TargetOp targetOp,
     llvm::OpenMPIRBuilder::TargetKernelRuntimeBounds &bounds) {
+  omp::LoopNestOp loopOp = castOrGetParentOfType<omp::LoopNestOp>(
+      targetOp.getInnermostCapturedOmpOp());
+  unsigned numLoops = loopOp ? loopOp.getNumLoops() : 0;
+
+  Value numThreads, numTeamsLower, numTeamsUpper, teamsThreadLimit;
+  llvm::SmallVector<Value> lowerBounds(numLoops), upperBounds(numLoops),
+      steps(numLoops);
+  extractHostEvalClauses(targetOp.getHostEvalVars(),
+                         llvm::cast<omp::BlockArgOpenMPOpInterface>(*targetOp)
+                             .getHostEvalBlockArgs(),
+                         numThreads, numTeamsLower, numTeamsUpper,
+                         teamsThreadLimit, &lowerBounds, &upperBounds, &steps);
+
   // TODO Handle IF clauses.
-  if (Value numTeamsLower = targetOp.getNumTeamsLower())
+  llvm::Value *&llvmTargetThreadLimit =
+      bounds.TargetThreadLimit.emplace_back(nullptr);
+  if (Value targetThreadLimit = targetOp.getThreadLimit())
+    llvmTargetThreadLimit = moduleTranslation.lookupValue(targetThreadLimit);
+
+  if (numTeamsLower)
     bounds.MinTeams = moduleTranslation.lookupValue(numTeamsLower);
 
   llvm::Value *&llvmMaxTeams = bounds.MaxTeams.emplace_back(nullptr);
-  if (Value numTeamsUpper = targetOp.getNumTeamsUpper())
+  if (numTeamsUpper)
     llvmMaxTeams = moduleTranslation.lookupValue(numTeamsUpper);
 
   llvm::Value *&llvmTeamsThreadLimit =
       bounds.TeamsThreadLimit.emplace_back(nullptr);
-  if (Value teamsThreadLimit = targetOp.getTeamsThreadLimit())
+  if (teamsThreadLimit)
     llvmTeamsThreadLimit = moduleTranslation.lookupValue(teamsThreadLimit);
 
-  if (Value numThreads = targetOp.getNumThreads())
+  if (numThreads)
     bounds.MaxThreads = moduleTranslation.lookupValue(numThreads);
 
-  if (Value tripCount = targetOp.getTripCount())
-    bounds.LoopTripCount = moduleTranslation.lookupValue(tripCount);
+  // To calculate the trip count, we first create a placeholder set of canonical
+  // loops which we then collapse. Then, we skip execution of the collapsed loop
+  // and remove the basic blocks that originally defined it. The trip count
+  // remains available in the entry block.
+  // TODO: Improve implementation by extracting the logic to calculate the trip
```

Comment:
Pardon my ignorance, but is there anything preventing us from doing it this way right now?

Reply:
I wanted to avoid touching the implementation of canonical loops creation, since I wasn't too familiar with it, but this approach was pretty terrible. I just made the change, which doesn't change much of the canonical loop creation, so hopefully should be fine.